Fix failures when default is flipped for weights_only

mikaylagawarecki opened the pull request:
Tests on XLA shard not fixed yet but there is an issue here https://github.com/pytorch/xla/issues/7799

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127627
* #132349



cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

mikaylagawarecki committed:
Update

[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot rebase

mikaylagawarecki committed:
Update

[ghstack-poisoned]

mikaylagawarecki committed:
Update

[ghstack-poisoned]

mikaylagawarecki committed:
Update

[ghstack-poisoned]

mikaylagawarecki committed:
Update

[ghstack-poisoned]

mikaylagawarecki committed:
Update

[ghstack-poisoned]

mikaylagawarecki commented on torch/fx/graph_module.py:
```diff
@@ -568,7 +568,8 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
                 module_repr = module.__repr__().replace("\r", " ").replace("\n", " ")
-                module_str = f"torch.load(r'{module_file}') # {module_repr}"
+                # weights_only=False as this is legacy code that saves the model
+                module_str = f"torch.load(r'{module_file}', weights_only=False) # {module_repr}"
```
@Chillee Wondering whether there is any BC concern here (tagging you as I noticed you added this API in https://github.com/pytorch/pytorch/pull/47544).

For context --  we are planning to flip the default for `torch.load` from `weights_only=False` --> `weights_only=True` in a future release. This PR is checking what will fail in existing CI and it seems this call needs `weights_only=False` as it serializes loads the serialized module.

Do you think there are existing modules dumped with `to_folder` that might break? 

I notice that above this says `@is_backward_compatible(False)`, so are we ok with breaking these?

mikaylagawarecki replied:
From offline discussion -- this is ok

mikaylagawarecki committed:
Update on "[Testing only] Flip default on weights_only"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix failures when default is flipped for weights_only"




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

albanD commented on torch/_weights_only_unpickler.py:
```diff
@@ -144,6 +144,8 @@ def _tensor_rebuild_functions():
         torch._utils._rebuild_nested_tensor,
         torch._utils._rebuild_wrapper_subclass,
+        # Allowlisting this, but not allowlisting the numpy functions by default
+        torch._utils._rebuild_device_tensor_from_numpy,
```
Is this because this function is always safe and help when a user is trying to allow numpy-based serialization?
I think we could be clearer on the intent here

mikaylagawarecki replied:
I think we should allowlist functions provided in `torch._utils` for rebuild, but as we discussed I don't think we want to allowlist the numpy functions themselves

mikaylagawarecki commented on test/test_serialization.py:
```diff
@@ -1140,7 +1152,7 @@ def test_weights_only_safe_globals_blocklist(self):
             f.seek(0)
             # safe_globals doesn't work even with allowlist
-            with torch.serialization.safe_globals([os.execv]):
+            with safe_globals([os.execv]):
```
It is, this tests that the `safe_globals` context manager isn't able to allowlist it

mikaylagawarecki replied:
we could, I'll open another PR to do this

albanD replied:
Sounds like a good small "actionable" issue to open for someone to jump on.

mikaylagawarecki commented on test/test_serialization.py:
```diff
@@ -880,25 +888,28 @@ def import_module(name, filename):
+                    self.assertEqual(len(w), 1)
+                    self.assertEqual(w[0].category, SourceChangeWarning)
+                    # self.assertEqual(w[0].category, FutureWarning)
+                    # self.assertEqual(w[1].category, SourceChangeWarning)
```
my bad

mikaylagawarecki commented on test/test_cpp_extensions_open_device_registration.py:
```diff
@@ -551,7 +556,18 @@ def test_open_device_numpy_serialization_map_location(self):
+                        np.dtype,
+                        _codecs.encode,
+                        type(np.dtype(np.float32))
+                        if np.__version__ < "1.25.0"
```
Hm search for "test_open_device_numpy_serialization_map_location" in https://ossci-raw-job-status.s3.amazonaws.com/log/28187737360 and you'll see the failure

albanD replied:
Ok! You can keep the code here as is.

FYI @atalman we have some CI that runs on older numpy which is unexpected?

mikaylagawarecki commented on test/quantization/bc/test_backward_compatibility.py:
```diff
@@ -113,7 +113,8 @@ def _test_op(
 
         input_tensor = torch.load(input_file)
-        qmodule.load_state_dict(torch.load(state_dict_file))
+        # weights_only = False as sometimes get ScriptObject here (weird)
```
weird as it was passing locally but not in CI, but I removed the weird

mikaylagawarecki committed:
Update on "Fix failures when default is flipped for weights_only"

Tests on XLA shard not fixed yet but there is an issue here https://github.com/pytorch/xla/issues/7799




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix failures when default is flipped for weights_only"

Tests on XLA shard not fixed yet but there is an issue here https://github.com/pytorch/xla/issues/7799




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

albanD reviewed (approved):
Ok!
Thanks for the update

mikaylagawarecki commented:
@pytorchbot merge -r

mikaylagawarecki committed:
Update

[ghstack-poisoned]
